cypress/integration/core_opensearch_dashboards/opensearch_dashboards/apps/query_enhancements/field_display_filtering.spec.js

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { MiscUtils } from '@opensearch-dashboards-test/opensearch-dashboards-test-library';
+import { DataExplorerPage } from '../../../../../utils/dashboards/data_explorer/data_explorer_page.po.js';
+import {
+  INDEX_CLUSTER_NAME,
+  INDEX_NAME,
+  INDEX_PATTERN_NAME,
+  SEARCH_ABSOLUTE_START_DATE,
+  SEARCH_ABSOLUTE_END_DATE,
+} from '../../../../constants.js';
+
+const miscUtils = new MiscUtils(cy);
+
+function selectDataSet(datasetType, language) {
+  switch (datasetType) {
+    case 'index':
+      DataExplorerPage.selectIndexDataset(INDEX_CLUSTER_NAME, INDEX_NAME, language);
+      break;
+    case 'index_pattern':
+      DataExplorerPage.selectIndexPatternDataset(INDEX_PATTERN_NAME, language);
+      if (language !== 'OpenSearch SQL') {
+        cy.setSearchAbsoluteDateRange(SEARCH_ABSOLUTE_START_DATE, SEARCH_ABSOLUTE_END_DATE);
+      }
+      break;
+  }
+}
+
+function checkTableFieldFilterActions(datasetType, language, isEnabled) {
+  selectDataSet(datasetType, language);
+
+  DataExplorerPage.getDiscoverQueryHits().should('not.exist'); // To ensure it waits until a full table is loaded into the DOM, instead of a bug where table only has 1 hit.
+
+  DataExplorerPage.checkDocTableFirstFieldFilterForAndOutButton(isEnabled);
+
+  if (isEnabled) {
+    DataExplorerPage.checkDocTableFirstFieldFilterForButtonFiltersCorrectField();
+    DataExplorerPage.checkDocTableFirstFieldFilterOutButtonFiltersCorrectField();
+  }
+}
+
+function checkExpandedTableFilterActions(datasetType, language, isEnabled) {
+  selectDataSet(datasetType, language);
+
+  DataExplorerPage.getDiscoverQueryHits().should('not.exist'); // To ensure it waits until a full table is loaded into the DOM, instead of a bug where table only has 1 hit.
+  DataExplorerPage.toggleDocTableRow(0);
+  DataExplorerPage.checkDocTableFirstExpandedFieldFirstRowFilterForFilterOutExistsFilterButtons(
+    isEnabled
+  );
+  DataExplorerPage.checkDocTableFirstExpandedFieldFirstRowToggleColumnButtonHasIntendedBehavior();
+
+  if (isEnabled) {
+    DataExplorerPage.checkDocTableFirstExpandedFieldFirstRowFilterForButtonFiltersCorrectField();
+    DataExplorerPage.checkDocTableFirstExpandedFieldFirstRowFilterOutButtonFiltersCorrectField();
+    DataExplorerPage.checkDocTableFirstExpandedFieldFirstRowExistsFilterButtonFiltersCorrectField();
+  }
+}
+
+describe('filter for value spec', () => {
+  beforeEach(() => {
+    cy.localLogin(Cypress.env('username'), Cypress.env('password'));
+    miscUtils.visitPage('app/data-explorer/discover');
+    cy.getNewSearchButton().click();
+  });
+  describe('filter actions in table field', () => {
+    describe('index pattern dataset', () => {
+      // filter actions should exist for DQL
```

Comment:
I would suggest we create a configs for dataset in the constants
```
export const DATASET_CONFIGS = {
  index_pattern: {
    type: 'Index Patterns',
    name: INDEX_PATTERN_NAME,
    languages: [
      { name: 'DQL', enabled: true },
      { name: 'Lucene', enabled: true },
      { name: 'OpenSearch SQL', enabled: false },
      { name: 'PPL', enabled: false }
    ]
  },
  index: {
    type: 'Indexes',
    clusterName: INDEX_CLUSTER_NAME,
    name: INDEX_NAME,
    languages: [
      { name: 'OpenSearch SQL', enabled: false },
      { name: 'PPL', enabled: false }
    ]
  }
};

```

Then for the tests like single lines or highly repeated, we could iterate
```
describe('filter for value spec', () => {
  beforeEach(() => {
    create workspace
    create index pattern
    cy.visitPage('app/data-explorer/discover');
    cy.getNewSearchButton().click();
  });

  // Use single test structure for both types
  const testCases = [
    { name: 'table field', verifyFn: verifyTableFilters },
    { name: 'expanded table', verifyFn: verifyExpandedFilters }
  ];

  testCases.forEach(({ name, verifyFn }) => {
    describe(`filter actions in ${name}`, () => {
      Object.entries(DATASET_CONFIGS).forEach(([type, config]) => {
        describe(`${type} dataset`, () => {
          config.languages.forEach(({ name: language, enabled }) => {
            it(`${language}`, () => {
              selectDataset(config, language);
              verifyFn(enabled);
            });
          });
        });
      });
    });
  });
});
```

Reply:
I like this approach. I will try something similar in my test suites, but one fear i have (that i do not have enough knowledge on yet to fully know) is that the `it` statement within the nesting might need to many `if` branches if the different permutations require special handlings. This has the potential to make the tests very brittle and when we have to refactor in the future, we might end up playing a lot of whack a mole. So I will see how it works on my test suites and report my findings afterwards